distribution/tools/plugin-cli/src/main/java/org/elasticsearch/plugins/InstallPluginCommand.java

```diff
@@ -206,24 +206,50 @@ protected void printAdditionalHelp(Terminal terminal) {
 
     @Override
     protected void execute(Terminal terminal, OptionSet options, Environment env) throws Exception {
-        String pluginId = arguments.value(options);
+        List<String> pluginId = arguments.values(options);
         final boolean isBatch = options.has(batchOption);
         execute(terminal, pluginId, isBatch, env);
     }
 
     // pkg private for testing
-    void execute(Terminal terminal, String pluginId, boolean isBatch, Environment env) throws Exception {
-        if (pluginId == null) {
-            throw new UserException(ExitCodes.USAGE, "plugin id is required");
+    void execute(Terminal terminal, List<String> pluginIds, boolean isBatch, Environment env) throws Exception {
+        if (pluginIds.isEmpty()) {
+            throw new UserException(ExitCodes.USAGE, "plugin ids are required");
         }
 
-        if ("x-pack".equals(pluginId)) {
-            handleInstallXPack(buildFlavor());
+        final Set<String> uniquePluginIds = new HashSet<>();
+        for (final String pluginId : pluginIds) {
+            if (uniquePluginIds.add(pluginId) == false) {
+                throw new UserException(ExitCodes.USAGE, "duplicate plugin id [" + pluginId + "]");
+            }
+        }
+
+        final List<Path> deleteOnFailure = new ArrayList<>();
+        final Set<PluginInfo> pluginInfos = new HashSet<>();
+        for (final String pluginId : pluginIds) {
+            try {
+                if ("x-pack".equals(pluginId)) {
+                    handleInstallXPack(buildFlavor());
+                }
+
+                final Path pluginZip = download(terminal, pluginId, env.tmpFile(), isBatch);
+                final Path extractedZip = unzip(pluginZip, env.pluginsFile());
+                deleteOnFailure.add(extractedZip);
+                final PluginInfo pluginInfo = installPlugin(terminal, isBatch, extractedZip, env, deleteOnFailure);
+                pluginInfos.add(pluginInfo);
+            } catch (final Exception installProblem) {
+                try {
+                    IOUtils.rm(deleteOnFailure.toArray(new Path[0]));
+                } catch (final IOException exceptionWhileRemovingFiles) {
+                    installProblem.addSuppressed(exceptionWhileRemovingFiles);
+                }
+                throw installProblem;
+            }
         }
 
-        Path pluginZip = download(terminal, pluginId, env.tmpFile(), isBatch);
-        Path extractedZip = unzip(pluginZip, env.pluginsFile());
-        install(terminal, isBatch, extractedZip, env);
+        for (final PluginInfo pluginInfo : pluginInfos) {
+            terminal.println("-> Installed " + pluginInfo.getName());
```

Comment:
Maybe we should emit each as we go? Dumping at the end seems counter to giving progress, which is the intention of this message IMO.

Reply:
I’m not sure we should dump as we go since it’s transactional? That is, “installed X” is a lie if we remove it if Y failed?

Reply:
Relates #51001